chalk/subdiagram.py

```diff
@@ -20,27 +22,39 @@
 
 
 Ident = Affine.identity()
-Subdiagram = Tuple[Diagram, Affine]
+Name = Tuple[str, ...]
```

Comment:
I have been using non-string names. Maybe we can make named() take Any but then wrap in a `Name` dataclass?

Reply:
I guess the question is do we expose Name to the user. Haskell does, but it forces you to use .> to build compound names.

Reply:
Sure, I will use more generic types for the atomic names (in principle, I think that anything that's hashable should work, since we index the subdiagram dictionary by a `Name`)?

Regarding the second point, do you have any advice? I've tried to keep things simple and allow the user to either specify an atomic name (a string in the initial version of the code) or a compound / qualified name (a tuple of strings); but I'm open to suggestions.

Reply:
Hmm. this is actually a little tricky. I think a tuple should work for qualification. But I have been using Tuples for just standard names -> i.e. (i,j) for a cell in a matrix. The question is will that break?

Reply:
Yes, you are right, I think my approach wouldn't have worked for that case.

As a second version, I was thinking about something along these lines:
```python
AtomicName = Any
Name = Tuple[AtomicName, ...]

class ApplyName(BaseDiagram):
    dname: Name
    diagram: Diagram

def named(self, *name: AtomicName) -> Diagram:
    return ApplyName(name, self)

def qualify(self, *name: AtomicName) -> Diagram:
    return self.accept(Qualify(name))

def get_subdiagram(self: Diagram, *name: AtomicName) -> Optional[Subdiagram]:
    return self.accept(GetSubdiagram(name), t=Ident)

def with_names(
    self: Diagram,
    names: List[Name],
    f: Callable[[List[Subdiagram], Diagram], Diagram],
) -> Diagram:
   ...
```
which would allow for something like this
```python
s = square(1)
squares = (
    (s.named("N", "W") | s.named("N", "E")) /
    (s.named("S", "W") | s.named("S", "E")))
squares = squares.qualify(0)
squares.get_subdiagram(0, "N", "W")
```
But unfortunately the API for `with_names` and `connect` get confusing since the name arguments have to be `Tuple`s and not `AtomicName`s (which in essence means that the user needs to know that the `Name`s are represented as `Tuple`s internally...):
```
root = circle(1).named("root")
leaves = hcat([circle(1).named(c) for c in "abcde"], sep=0.5).center()
nodes = root / vstrut(2) / leaves

def connect(subs, nodes):
    ...

for c in "abcde":
    nodes = nodes.with_names([("root", ), (c, )], connect)
```

Reply:
Yes, I agree this a bit confusing. My intuition is that we should have the correct, verbose way first. There are a couple areas in the code that we may want "abbreviations" as you have mentioned in the past (colors, P2s, etc). Maybe we just have ways to make it less verbose on top of this foundation.

Reply:
Thanks a lot for the feedback @srush! I've updated the API to be more explicit: all naming functions (`named`, `qualify`, `with_names`, `get_subdiagram`, etc.) are now expected to take as input an instance of the `Name` class. There are some examples in the [`examples/subdiagrams.py`](https://github.com/chalk-diagrams/chalk/blob/8e4e5ab738450aee41257d1b4bdee460204a064f/examples/subdiagrams.py) file. When you get a chance, please let me know what you think of the current implementation. I'll update the other examples (`lenet.py`, `lattice.py`, `tree.py`, etc.) and the documentation, once we converge to a solution.